### ISSUE TITLE
vAmiga crash with drag&drop

### ISSUE THREAD

**Issue by Gianmarco72** [dirkwhoffmann/vAmiga#435]
Sometimes when I drag and drop a rom file onto the vAmiga window it crashes.
I'm not sure but seems to happen more often when the rom file is in zip format.
<img width="1112" alt="Schermata 2020-11-04 alle 12 27 53" src="https://user-images.githubusercontent.com/45571833/98106540-b591e880-1e99-11eb-9163-51aac69815b1.png">
I also enclose the log file.
[vAmiga crash log.txt](https://github.com/dirkwhoffmann/vAmiga/files/5487489/vAmiga.crash.log.txt)

**Comment by dirkwhoffmann** [comment]
From your description I deduce it's indeterministic behavior. I guess you opened the Rom config panel and dropped the Rom file into one of the drop target fields there. Right? Was the emulator running when the crash happened? I am asking because I need to find a way to reproduce the problem on my machine.

**Comment by Gianmarco72** [comment]
I'm sorry, I try to explain better. When I wrote ROM I meant an ADF file (e.g. of a game).
This is what happen: I start vAmiga then I drag an ADF to the main window and sometimes vAmiga crashes immediately.
I'm not sure but it seems to happen more often with files in ZIP format.

**Comment by Gianmarco72** [comment]
Update: today I tried to open the "Configure Virtual Machine" panel and vAmiga crashed exactly like the drag and drop function.🤔

**Comment by dirkwhoffmann** [comment]
I got a crash here. Maybe it's related: 

<img width="1519" alt="Bildschirmfoto 2020-11-25 um 16 03 27" src="https://user-images.githubusercontent.com/12561945/100497310-e1cc1c80-315a-11eb-9a55-1ec643c1e6f9.png">

**Comment by Gianmarco72** [comment]
Meantime in the last two weeks vAmiga has crashed a couple of times apparently randomly, even with version 0.9.15.🤯

**Comment by dirkwhoffmann** [comment]
There was a big bug in the audio stream ring buffer class. The buffer is written by one thread and read by another. Every first year student knows that a mutex is required in such cases to synchronize access, but there was no such mutex 🙄. The code is from VirtualC64 which means that VirtualC64 is missing this mutex for about 15 years now. I better not tell anyone 😬.

The mutex is part of v0.9.15.1.

I'll close this issue in the hope that it fixes the crash this thread is about. Please reopen if it doesn't fix it.

**Comment by Gianmarco72** [comment]
Nope... crash again with v0.9.15.1 😕
<img width="1112" alt="Schermata 2020-12-06 alle 17 04 26" src="https://user-images.githubusercontent.com/45571833/101285421-67983980-37e5-11eb-9d25-29a602b73603.png">
 
I dragged an ADF (in ZIP format) on the main screen of vAmiga, then the "choose drives" window appeared
<img width="942" alt="Schermata 2020-12-06 alle 17 10 13" src="https://user-images.githubusercontent.com/45571833/101285588-569bf800-37e6-11eb-9fd8-419c84ddf29a.png">
I clicked on "Cancel" ... and BOOM, crash! 😖

**Comment by dirkwhoffmann** [comment]
Strange. The last thing it does from within my code is this: 

```
5   com.apple.AMDMTLBronzeDriver  	0x00007fff2437ba60 -[BronzeMtlTexture replaceRegion:mipmapLevel:withBytes:bytesPerRow:] + 578
6   dirkwhoffmann.vAmiga          	0x000000010a4ccb8e 0x10a346000 + 1600398
```

`replaceRegion` replaces a portion of the texture with new data from the emulator.  I can't imagine how this interacts with the use case "dragging in an ADF" at the moment. 🤔

**Comment by dirkwhoffmann** [comment]
By removing all warnings reported by the clang Sanitizer in v0.9.16.1, I've found a severe memory bug in the setup of the Merge shader (the old code wrote beyond the bounds of an array). With some luck, this fixed the crash you are experiencing.

**Comment by Gianmarco72** [comment]
Ok, I'll try the new version 0.9.16.1 🤞🏻

**Comment by dirkwhoffmann** [comment]
I'll close this in the hope that it's fixed in 0.9.16.1 (or 0.9.16.2). Please reopen if not.

**Comment by dirkwhoffmann** [comment]
This time, I got the very same error myself: 

<img width="1529" alt="Bildschirmfoto 2021-01-06 um 11 39 29" src="https://user-images.githubusercontent.com/12561945/103759456-0cd84480-5014-11eb-9b0f-10348dea7220.png">

<img width="1529" alt="Bildschirmfoto 2021-01-06 um 11 33 33" src="https://user-images.githubusercontent.com/12561945/103759448-0ba71780-5014-11eb-8d6b-34563e3ae943.png">

What I did is to work with multiple emulator instances at the same time. The log shows that the crash happened during the initialisation of the second instance (right after makeTouchBar()). I guess the textures are not yet created at this point in time. Since the bug never occurred on my machine when I was working with a single emulator instance, I expect some asynchronous stuff going on during initialisation. The easiest explanation would be that drawing starts before initialisation is completed.

**Comment by Gianmarco72** [comment]
I usually work with a single instance and the crash still occurs.
Question for you: maybe, using drag & drop to boot a ROM, a second instance starts automatically?🤔

**Comment by dirkwhoffmann** [comment]
I haven't managed to reproduce it a second time yet, but I noticed something. This is what you've been doing:

> I clicked on "Cancel" ... and BOOM, crash! 😖

According to my log, I seem to have done the same. 

For reference: This is what's happening when hitting "Cancel":

 ```
   @IBAction override func cancelAction(_ sender: Any!) {
         
         track()
                         
         hideSheet()
         try? myDocument.persistScreenshots()
     }

    func hideSheet() {
    
        if let win = window {
            win.orderOut(self)
            parent.window?.endSheet(win, returnCode: .cancel)
        }
    }
```

**Comment by dirkwhoffmann** [comment]
I didn't experience the crash in one of the later releases any more, so I'll close the issue in the hope that it is gone for good. Please reopen the issue if it is not. The main problem here is to make the crash reproducible in some way.

**Comment by dirkwhoffmann** [comment]
I think I finally fixed this with commit e0bffef. At the end, it was something very simple: The GPU texture is larger than it's counterpart in CPU memory and I copied as many bytes as the GPU texture comprises. In most cases this doesn't do any harm, because the CPU reads from "invalid" memory and doesn't write to it. However, in rare cases, this can cause a segmentation fault, because there is no guarantee that the CPU has the permission to read all addressed memory cells.
